### PR TITLE
Allow to programmatically register additional source folders from CompilationProvider

### PR DESCRIPTION
**Pull request by mariofusco**
…pilationProvider

### CODE REVIEW

File: core/devmode/src/main/java/io/quarkus/dev/CompilationProvider.java [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
 
 import java.io.File;
+import java.util.Collections;
 import java.util.Set;
 
 public interface CompilationProvider {
 
     Set<String> handledExtensions();
 
+    default Set<String> handleSourcePaths() {
```
Comment (gastaldi) [inline core/devmode/src/main/java/io/quarkus/dev/CompilationProvider.java]:
```suggestion
    default Set<String> handledSourcePaths() {
```

File: core/devmode/src/main/java/io/quarkus/dev/RuntimeCompilationSetup.java [hunk 2]
```diff
@@ -32,6 +32,10 @@ public static RuntimeUpdatesProcessor setup(DevModeContext context) throws Excep
         if (!context.getModules().isEmpty()) {
             ServiceLoader<CompilationProvider> serviceLoader = ServiceLoader.load(CompilationProvider.class);
             List<CompilationProvider> compilationProviders = new ArrayList<>();
+            for (CompilationProvider provider : serviceLoader) {
+                compilationProviders.add(provider);
+                context.getModules().forEach(moduleInfo -> moduleInfo.addSourcePaths(provider.handleSourcePaths()));
```
Comment (gastaldi) [inline core/devmode/src/main/java/io/quarkus/dev/RuntimeCompilationSetup.java]:
```suggestion
                context.getModules().forEach(moduleInfo -> moduleInfo.addSourcePaths(provider.handledSourcePaths()));
```

File: core/devmode/src/main/java/io/quarkus/dev/RuntimeCompilationSetup.java [hunk 3]
```diff
@@ -32,6 +32,10 @@ public static RuntimeUpdatesProcessor setup(DevModeContext context) throws Excep
         if (!context.getModules().isEmpty()) {
             ServiceLoader<CompilationProvider> serviceLoader = ServiceLoader.load(CompilationProvider.class);
             List<CompilationProvider> compilationProviders = new ArrayList<>();
+            for (CompilationProvider provider : serviceLoader) {
+                compilationProviders.add(provider);
```
Comment (stuartwdouglas) [inline core/devmode/src/main/java/io/quarkus/dev/RuntimeCompilationSetup.java]: This looks like the compilation providers are being added twice, as the line below the loop also adds them

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

File: core/devmode/src/main/java/io/quarkus/dev/RuntimeCompilationSetup.java (hunk 3 above)
Comment (mariofusco) [reply core/devmode/src/main/java/io/quarkus/dev/RuntimeCompilationSetup.java]: You're right, I fixed it by removing that line at the end of loop. Thanks for having found this.

**Comment by mariofusco** [comment]
If this pull request is ok now I'd appreciate if this could be part of the next quarkus release. It is necessary to finalize our drools/jbpm extension for quarkus.

**Comment by mkouba** [comment]
LGTM. @stuartwdouglas any comments?

**Review by stuartwdouglas — APPROVED** [review]

**Pull request merged**